### ISSUE TITLE
Multiprocessing backed parallel loops cannot be nested below threads, setting n_jobs=1

### ISSUE THREAD

**Issue by AlexanderKUA** [joblib/joblib#180]
What is the reason of such issue in joblib? `'Multiprocessing backed parallel loops cannot be nested below threads, setting n_jobs=1'`
What should I do to avoid such issue?

Actually I need to implement XMLRPC server which run heavy computation in background thread and report current progress through polling from UI client. It uses scikit-learn which are based on joblib.

P.S.: I've simply changed name of the thread to "MainThread" to avoid such warning and everything looks working good (run in parallel as expected without issues). What might be a problem in future for such workaround?

**Comment by ogrisel** [comment]
It always works even we you get, it's just that nested calls are not supported by multiprocessing, therefore joblib detects that and convert nested calls into sequential code on the fly and let you know that this happens.

So the warning is not a bug, it's a feature. Closing. Feel free to reopen if you think you get the warning in cases where you don't expect it to happen (e.g. not a nested parallelism case).

**Comment by rth** [comment]
While using multiprocessing below threads is not supported, the current [test to detect it](https://github.com/joblib/joblib/blob/master/joblib/_parallel_backends.py#L280) relies on the thread name being equal to `MainThread` which can produce false positives. 

For instance, I'm trying  to use joblib parallel inside uWSGI (in a non multi-threaded setup). Different workers use a single thread, but the default thread name is something of the form `uWSGIWorker2Core0`, and renaming the thread name to `MainThread` appears to make everything work fine (same as for the OP), though I hope this doesn't break anything in uWSGI.

As suggested in this [SO answer](https://stackoverflow.com/a/23207116/1791279), a more robust testing for multiprocessing nested under threads in joblib could be,

    isinstance(threading.current_thread(), threading._MainThread))

instead of the currently used,

    threading.current_thread().name == 'MainThread'

For my uWSGI setup, the latter evaluates to `False`, while the former is `True`.

**Comment by lesteve** [comment]
Seems sensible, PR welcome. A test with a Parallel in the main thread whose name is different than MainThread would be great, if that's not too painful to do.

**Comment by ogrisel** [comment]
+1

**Comment by ogrisel** [comment]
@rth let me do the PR for you.

**Comment by rth** [comment]
@ogrisel  Sure, thanks! Sorry meant to look into it, but never got to it.

**Comment by ogrisel** [comment]
Done here: #414. @rth I would love your feedback in case it does not fix your particular issue with uWSGI as expected.

**Comment by wangyexiang** [comment]
> I've simply changed name of the thread to "MainThread" to avoid such warning and everything looks working good (run in parallel as expected without issues). 
Dear sir. How to chang name of the thread to "MainThread"? Could you give me a example? @AlexanderKUA

**Comment by AlexanderKUA** [comment]
`threading.current_thread().setName('MainThread')`

**Comment by rth** [comment]
This was fixed a long time ago and renaming the thread to `MainThread` should no longer be necessary. Please open a new issue of you still encounter difficulties with this.
